src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.TypeInit.cs

```diff
@@ -26,7 +33,7 @@ partial class CompilerTypeSystemContext
         /// </summary>
         public bool HasLazyStaticConstructor(TypeDesc type)
         {
-            return type.HasStaticConstructor && !HasEagerConstructorAttribute(type);
+            return type.HasStaticConstructor && !HasEagerConstructorAttribute(type) && !HasExplicitClassConstructor(type);
```

Comment:
I tried to make the additional condition cheap. This might go away though in a follow-up if we go the Initialize() route.